File: bench/bench_profile.py
#!/opt/anaconda3/bin/python
# bench/bench_profile.py
# Single-library, single-size, N-iteration profiling driver for NumPy and
# PyTorch. Output format mirrors bench_profile.cpp / SMETest::profile() so
# profile.sh and profile_power.sh can parse it the same way.
#
# Usage:
#   bench_profile.py <numpy|pytorch> <M> <K> <N> <iters>

import os
import sys
import time

# Pin to single thread BEFORE importing numpy/torch.
os.environ["OMP_NUM_THREADS"]        = "1"
os.environ["MKL_NUM_THREADS"]        = "1"
os.environ["OPENBLAS_NUM_THREADS"]   = "1"
os.environ["VECLIB_MAXIMUM_THREADS"] = "1"
os.environ["NUMEXPR_NUM_THREADS"]    = "1"

import numpy as np

if len(sys.argv) != 6:
    sys.exit(f"usage: {sys.argv[0]} <numpy|pytorch> M K N iters")

lib = sys.argv[1]
M, K, N, iters = (int(x) for x in sys.argv[2:6])

rng = np.random.default_rng(42)
A_np = rng.random((M, K), dtype=np.float32)
B_np = rng.random((K, N), dtype=np.float32)

if lib == "numpy":
    label = "numpy"
    matmul = lambda: np.matmul(A_np, B_np)
elif lib == "pytorch":
    import torch
    torch.set_num_threads(1)
    torch.set_num_interop_threads(1)
    label = "pytorch"
    A_pt = torch.from_numpy(A_np)
    B_pt = torch.from_numpy(B_np)
    matmul = lambda: torch.matmul(A_pt, B_pt)
else:
    sys.exit(f"unknown library: {lib}  (expected numpy|pytorch)")

print("  Threads: 1")

matmul()  # warmup
t0 = time.perf_counter()
for _ in range(iters):
    matmul()
sec = time.perf_counter() - t0

gflops      = (2.0 * M * N * K * iters / 1e9) / sec
ms_per_iter = sec * 1000.0 / iters
print(f"  [{label}] {M}x{K}x{N} iters={iters} : {gflops:.1f} GFLOPS ({ms_per_iter:.2f} ms/iter)")

File: bench/bench_profile.cpp
// bench/bench_profile.cpp
// Single-library, single-size, N-iteration driver for Instruments PMU /
// powermetrics profiling. Mirrors the output format of SMETest::profile()
// in sme/test_sme.cpp so scripts/profile.sh and scripts/profile_power.sh can reuse the
// same parsers (looks for "GFLOPS" / "iters=" / "Threads:" lines).
//
// Usage:
//   bench_profile <accel|oblas> <M> <K> <N> <iters>
//
// Env: forces single-thread for both Accelerate (vecLib) and OpenBLAS.

#include <iostream>
#include <iomanip>
#include <vector>
#include <chrono>
#include <string>
#include <cstdlib>
#include <random>
#include <functional>

#include <Accelerate/Accelerate.h>
#include <dlfcn.h>
#include <omp.h>

using Clock = std::chrono::high_resolution_clock;
using Ms    = std::chrono::duration<double, std::milli>;

static void fill_random(std::vector<float>& v) {
    std::mt19937 gen(42);
    std::uniform_real_distribution<float> dis(-1.0f, 1.0f);
    for (auto& x : v) x = dis(gen);
}

static void accel_sgemm(const float* A, const float* B, float* C,
                        int M, int N, int K) {
    cblas_sgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans,
                M, N, K, 1.0f, A, K, B, N, 0.0f, C, N);
}

using cblas_sgemm_fn = void (*)(int, int, int, int, int, int,
                                float, const float*, int,
                                const float*, int, float, float*, int);
static cblas_sgemm_fn g_openblas_sgemm = nullptr;

// Try $OPENBLAS_DYLIB first, then the dynamic loader's own search path, then
// the usual Homebrew prefixes (Apple Silicon, then Intel).
static void* dlopen_openblas() {
    const char* env = std::getenv("OPENBLAS_DYLIB");
    const char* candidates[] = {
        env,
        "libopenblas.dylib",
        "/opt/homebrew/opt/openblas/lib/libopenblas.dylib",
        "/usr/local/opt/openblas/lib/libopenblas.dylib",
    };
    for (const char* path : candidates) {
        if (!path) continue;
        if (void* h = dlopen(path, RTLD_NOW | RTLD_LOCAL)) return h;
    }
    return nullptr;
}

static bool load_openblas() {
    void* handle = dlopen_openblas();
    if (!handle) {
        std::cerr << "  [error] could not load libopenblas.dylib: " << dlerror() << "\n"
                  << "          install with 'brew install openblas', or set "
                     "OPENBLAS_DYLIB=/path/to/libopenblas.dylib\n";
        return false;
    }
    g_openblas_sgemm = reinterpret_cast<cblas_sgemm_fn>(
        dlsym(handle, "cblas_sgemm"));
    if (!g_openblas_sgemm) {
        std::cerr << "  [error] dlsym cblas_sgemm: " << dlerror() << "\n";
        return false;
    }
    return true;
}

static void openblas_sgemm(const float* A, const float* B, float* C,
                           int M, int N, int K) {
    g_openblas_sgemm(101, 111, 111, M, N, K,
                     1.0f, A, K, B, N, 0.0f, C, N);
}

int main(int argc, char** argv) {
    if (argc != 6) {
        std::cerr << "usage: " << argv[0] << " <accel|oblas> M K N iters\n";
        return 1;
    }
    const std::string lib = argv[1];
    const int M = std::atoi(argv[2]);
    const int K = std::atoi(argv[3]);
    const int N = std::atoi(argv[4]);
    const int iters = std::atoi(argv[5]);

    omp_set_num_threads(1);
    setenv("VECLIB_MAXIMUM_THREADS", "1", 1);
    setenv("OPENBLAS_NUM_THREADS",   "1", 1);
    std::cout << "  Threads: " << omp_get_max_threads() << "\n";

    std::vector<float> A(static_cast<size_t>(M) * K),
                       B(static_cast<size_t>(K) * N),
                       C(static_cast<size_t>(M) * N);
    fill_random(A); fill_random(B);

    std::function<void()> fn;
    std::string label;
    if (lib == "accel") {
        label = "accel";
        fn = [&]{ accel_sgemm(A.data(), B.data(), C.data(), M, N, K); };
    } else if (lib == "oblas") {
        if (!load_openblas()) return 1;
        label = "oblas";
        fn = [&]{ openblas_sgemm(A.data(), B.data(), C.data(), M, N, K); };
    } else {
        std::cerr << "unknown library: " << lib << "  (expected accel|oblas)\n";
        return 1;
    }

    fn(); // warmup
    auto t0 = Clock::now();
    for (int i = 0; i < iters; ++i) fn();
    double sec = Ms(Clock::now() - t0).count() / 1000.0;

    double gflops      = (2.0 * double(M) * double(N) * double(K) * iters / 1e9) / sec;
    double ms_per_iter = sec * 1000.0 / iters;
    std::cout << "  [" << label << "] " << M << "x" << K << "x" << N
              << " iters=" << iters
              << " : " << std::fixed << std::setprecision(1) << gflops << " GFLOPS ("
              << std::setprecision(2) << ms_per_iter << " ms/iter)\n";
    return 0;
}
